src/util/alloc/free_list_allocator.rs

```diff
@@ -0,0 +1,530 @@
+        if cell.is_zero() {
+            return cell; // return failed allocation
+        }
+        let next_cell = unsafe { cell.load::<Address>() };
+        // Clear the link
+        unsafe { cell.store::<Address>(Address::ZERO) };
+        debug_assert!(
+            next_cell.is_zero() || block.includes_address(next_cell),
+            "next_cell {} is not in {:?}",
+            next_cell,
+            block
+        );
+        block.store_free_list(next_cell);
+
+        // Zeroing memory right before we return it.
+        // If we move the zeroing to somewhere else, we need to clear the list link here: cell.store::<Address>(Address::ZERO)
+        let cell_size = block.load_block_cell_size();
+        crate::util::memory::zero(cell, cell_size);
+
+        // Make sure the memory is zeroed. This looks silly as we zero the cell right before this check.
+        // But we would need to move the zeroing to somewhere so we can do zeroing at a coarser grainularity.
+        #[cfg(debug_assertions)]
+        {
+            let mut cursor = cell;
+            while cursor < cell + cell_size {
+                debug_assert_eq!(unsafe { cursor.load::<usize>() }, 0);
+                cursor += crate::util::constants::BYTES_IN_ADDRESS;
+            }
+        }
+
+        cell
+    }
+
+    // Find an available block when stress GC is enabled. This includes getting a block from the space.
+    fn find_free_block_stress(&mut self, size: usize, align: usize) -> Block {
+        let mut block = Self::find_free_block_with(
+            &mut self.available_blocks_stress,
+            &mut self.consumed_blocks,
+            size,
+            align,
+        );
+        if block.is_zero() {
+            block = self.recycle_local_blocks(size, align, true);
+        }
+        if block.is_zero() {
+            block = self.acquire_fresh_block(size, align, true);
+        }
+        block
+    }
+
+    // Find an available block from local block lists
+    #[inline(always)]
+    fn find_free_block_local(&mut self, size: usize, align: usize) -> Block {
+        let block = Self::find_free_block_with(
+            &mut self.available_blocks,
+            &mut self.consumed_blocks,
+            size,
+            align,
+        );
+        if block.is_zero() {
+            self.recycle_local_blocks(size, align, false)
+        } else {
+            block
+        }
+    }
+
+    // Find an available block
+    // This will usually be the first block on the available list. If all available blocks are found
+    // to be full, other lists are searched
+    // This function allows different available block lists -- normal allocation uses self.avaialble_blocks, and precise stress test uses self.avialable_blocks_stress.
+    #[inline(always)]
+    fn find_free_block_with(
+        available_blocks: &mut BlockLists,
+        consumed_blocks: &mut BlockLists,
+        size: usize,
+        align: usize,
+    ) -> Block {
```

Comment:
I think we still need to take offset into consideration when searching for free block.

For example, if the VM requests an object of 64 bytes, it can be allocated into the 64-byte size class; but if the VM requests that the object must be aligned to 16 bytes at +4 byte offset, then the object has to be at least 64 + (16 - 4) = 76 bytes large, and it has to be allocated into the next size class, i.e. 80 bytes.

```suggestion
    fn find_free_block_with(
        available_blocks: &mut BlockLists,
        consumed_blocks: &mut BlockLists,
        size: usize,
        align: usize,
        offset: usize,
    ) -> Block {
```

Reply:
I think `offset` does not matter here. If I understand correct, `offset` means our allocation result `res` should have `(res + offset).is_aligned_to(align)`, and the allocation range is `[res, res + size)`.

`offset` is not passed to this method, as it is not required by `get_maximum_aligned_size()`.
https://github.com/mmtk/mmtk-core/blob/06144875903b36b432587eb7462e0272750b71d9/src/util/alloc/allocator.rs#L94-L98